chore: logging ADR (#2588)
zarf-dev/zarf: adr/0025-logging.md

# 25. Logging

Date: 2024-06-06

## Status

Proposed

## Context

> Zarf is currently using an in-house built logging solution which in turn depends on pterm. This solution is used to output information to the end user who is using the Zarf CLI. This output is for the most part formatted with the purpose of look good for users of the CLI. Logging function calls are done both in the CLI specific code as well as in the packages. The logging is implemented in such a way that different levels exist with different impacts and log destinations. A common pattern that is used is to call `message.Fatal` whenever an error occurs. This will output the message to STDERR while writing the actual error message to a debug log and exit the program. Exiting the program in this manner makes unit testing difficult while also skipping proper context handling skipping any clean-up that was intended to run before exiting the program. Some logging components like the progress bar and spinner are accessed through a shared global state that is not thread safe. The coupling to logging becomes complicated as disabling the progress bar is a challenge while multi threading these functions which access the global state results in complicated access patterns.

Reviewer: ```suggestion
Zarf is currently using an in-house built logging solution which in turn depends on pterm. This solution is used to output information to the end user who is using the Zarf CLI. This output is for the most part formatted with the purpose of CLI user expierience. Logging function calls are done both in the CLI specific code as well as in the packages. The logging is implemented in such a way that different levels exist with different impacts and log destinations. A common pattern that is used is to call `message.Fatal` whenever an error occurs. This will output the message to STDERR while writing the actual error message to a debug log and exiting the program. Exiting the program in this manner makes unit testing difficult while also skipping proper context handling and skipping any clean-up that was intended to run before exiting the program. Some logging components like the progress bar and spinner are accessed through a shared global state that is not thread safe. The coupling to logging becomes complicated as disabling the progress bar is a challenge while multi threading these functions which access the global state resulting in complicated access patterns.
```

## Decision

I am proposing to completely refactor the logging functionality to follow a more standardized format using the newish slog interfaces. On top of that we would refactor the current internationalization by converting them to standardized errors.

* Replace the message package with a slog interface.
* Replace the lang package with static errors.
* Remove any use of `message.Fatal` and instead return errors.
* Refactor use of `message.WarnErr` to either return error or to log the error.

> * Refactor existing functions that print formatted outputs to defer the printing to the CLI.

Reviewer 2: Anything that does fancy UI things should prob happen in CMD, all lib code should use the `slog` interface

Author: While this is true it is not currently possible right now. The spinner is too heavily integrated into the all of the code base. The best we can do initially is to remove the need for a global spinner that has threading issues. Then in the future we can look at how we can move UI like behavior into the cmd package.

Reviewer: It may not be the best pattern but an implementation to look into could be what was done with the ProgressWriter and Maru - https://github.com/defenseunicorns/maru-runner/blob/main/src/message/progress.go#L29 - that makes the spinner and progressbar a bit more generic and a bit less decoupled.

The main thing I don't love about that implementation though it that spinner's `Write()` and progressbar's `Write()` mean slightly different things, but it is probably a decent starting point for `* Define a new interface for spinner and progress bar that is passed in as a function parameter.`

> * Define a new interface for spinner and progress bar that is passed in as a function parameter.

Reviewer 3: I like these suggestions. Could we also aim to reduce the overall log output in Zarf as part of this refactor? What do you think?

Author: I dont think that is the explicit plan but it will probably happen on its own. We should revisit the logs after this change is implemented.

The message package currently exports the following functions which should be replaced by its logr counterpart.

| Function | Replacement | Comment |
| --- | --- | --- |
| ZarfCommand | Info | Just formats a slice as a command. |
| Command | Info | Outputs a command with a prefix and style. |

> | Debug | Degub | |

Reviewer 2: ```suggestion
| Debug | Debug | |
```

Reviewer 4: +1

| Debugf | Debug | |

> | ErrorWebf | N/A | Not used anywhere. |

Reviewer: (historical context - this is vestigial from when Zarf had a web UI)

| Warn | Warn | |
| Warnf | Warn | |
| WarnErr | Warn | |
| WarnErrf | Warn | |
| Fatal | N/A | Should not be used. |
| Fatalf | N/A | Should not be used. |
| Info | Info | |
| Infof | Info | |
| Success | Info | Seems to be a info log with a checkmark prefix. |
| Successf | Info | Seems to be a info log with a checkmark prefix. |
| Question | ? | Open question how to resolve this. |
| Questionf | ? | Open question how to resolve this. |
| Note | Info | Should just be an info or maybe a debug log. |
| Notef | Info | Should just be an info or maybe a debug log. |
| Title | ? | Not sure how this should be replaced as it formats with separator. |
| HeaderInfof | ? | |

> | HorizontalRule | ? | |

Reviewer: For these CLI UXy pieces it may be good to take a look at the Figma files @Madeline-UX put together to see how they fit / are organized there - there may be ways to build some of this in an "interactive" package that intentionally prints differently to the log vs the CLI - Horizontal Rules, ColorWrapped YAML, and Questions are really only there to be used in the CLI UI and could/should probably be represented differently in the log

| ColorWrap | ? | Should this be used? | 
| PrintConnectStringTable | N/A | This logic should not exist in the message package. |
| PrintCredentialTable | N/A | This logic should not exist in the message package. |
| PrintComponentCredential | N/A | This logic should not exist in the message package. |
| PrintCredentialUpdates | N/A | This logic should not exist in the message package. |
| Spinner | Interface | New Spinner interface. |
| ProgressBar | Interface | New progress bar interface. |

The majority of simple logging changes should be possible with little signature changes. Replacing the existing output with a slog interface would allow other implementations to be used. A challenge initially may be the changes to table output formatting to make it work properly. This change will require some refactoring of existing code. A requirement for the changes is that they have to improve the UX for users looking at log files. As I understand the present day the spinners will cause a new line everytime they update, resulting in a lot of bloat. A goal should be to make sure that this does not happen in the future.

> Spinners and progress bars however are a bit more challenging. They need to be refactored so that they are no long instantiated outside of the CLI code. They should instead be passed as a function parameter to each function that needs them. A good example of a project that solves the problem in a similar manner is Kind. In Kind they create a [status object from the logger](https://github.com/kubernetes-sigs/kind/blob/7799e72306db315ea4f4b1cac90ff68404da4f28/pkg/internal/cli/status.go#L39) which they then [pass to where it is needed](https://github.com/kubernetes-sigs/kind/blob/7799e72306db315ea4f4b1cac90ff68404da4f28/pkg/cluster/internal/create/create.go#L133). Doing so results in a single status object created which is reused where ever it is needed. A lot of inspiration can be take from how Kind deals with CLI output. While they use klog instead of slog there are a lot of similarities. They have for example a check if the output is in a terminal or not, and will disable the spinner accordingly. 

Reviewer: ```suggestion
Spinners and progress bars however are a bit more challenging. They need to be refactored so that they are no longer instantiated outside of the CLI code. They should instead be passed as a function parameter to each function that needs them. A good example of a project that solves the problem in a similar manner is Kind. In Kind they create a [status object from the logger](https://github.com/kubernetes-sigs/kind/blob/7799e72306db315ea4f4b1cac90ff68404da4f28/pkg/internal/cli/status.go#L39) which they then [pass to where it is needed](https://github.com/kubernetes-sigs/kind/blob/7799e72306db315ea4f4b1cac90ff68404da4f28/pkg/cluster/internal/create/create.go#L133). Doing so results in a single status object created which is reused where ever it is needed. A lot of inspiration can be take from how Kind deals with CLI output. While they use klog instead of slog there are a lot of similarities. They have for example a check if the output is in a terminal or not, and will disable the spinner accordingly. 
```

> Here is a suggestion for how a thread safe spinner could be implemented with a shared logger. This also allows for paralell spinners and progress bars.

Reviewer 3: Have you verified that this is thread-safe? Running `go run -race main.go` shows many data races in the pterm printers, likely related to issue #2558. It seems upstream changes to the pterm printers might be needed for thread safety

Author: I initially did not understand all of the race issues in pterm. After looking through things it is obvious that there are issues in pterm that we cannot fix outside of the project. We have two options. Either we wait for pterm to fix the issue internally, or we find and alternative to pterm that offer the same features. It is slightly difficult to find a project that supports multi writer ontop of spinners and progress bars that is not from charm.

```golang
package main

import (

> 	"context"

Reviewer 2: I'm a big fan of the `FromContext`/`WithContext` pattern https://pkg.go.dev/github.com/charmbracelet/log#FromContext https://pkg.go.dev/github.com/charmbracelet/log#WithContext

Reviewer 2: should make it a little easier / standarized on how the logger is passed around library code: its always in `ctx`, `FromContext` also nicely handles if no logger has been passed, thereby preventing nil pointer errors / erroneous printing.

## Decision

I am proposing to completely refactor the logging functionality to follow a more standardized format using the newish slog interfaces. On top of that we would refactor the current internationalization by converting them to standardized errors.

* Replace the message package with a slog interface.
* Replace the lang package with static errors.
* Remove any use of `message.Fatal` and instead return errors.
* Refactor use of `message.WarnErr` to either return error or to log the error.

> * Refactor existing functions that print formatted outputs to defer the printing to the CLI.

Author: You ask and I deliver.
